### PR TITLE
Porting WSDL related components from wso2-support/carbon-apimgt#2209 

### PR DESCRIPTION
**Pull request by Chamindu36**
### Purpose
1. Porting WSDL related components from wso2-support/carbon-apimgt#2209

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/wsdl/WSDL20ProcessorImpl.java [hunk 1]
```diff
@@ -254,6 +255,12 @@ public ErrorHandler getError() {
         return error;
     }
 
+    @Override
+    public void loadXSDs(APIMWSDLReader wsdlReader, String url) {
+        throw new UnsupportedOperationException("This method is not implemented");
+    }
+
```
Comment (hasuniea) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/wsdl/WSDL20ProcessorImpl.java:262]: Remove new lines.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/wsdl/WSDL11SOAPOperationExtractor.java [hunk 2]
```diff
@@ -245,8 +303,11 @@ private boolean initModels() {
                     WSDLParamDefinition wsdlParamDefinition = new WSDLParamDefinition();
                     ModelImpl model = new ModelImpl();
                     Property currentProperty = null;
-                    traverseTypeElement(node, null, model, currentProperty);
-                    if (StringUtils.isNotBlank(model.getName())) {
+                    try {
+                        traverseTypeElement(node, null, model, currentProperty);
+                    } catch (APIManagementException e) {
+                        throw new APIMgtWSDLException(e);
+                    }                    if (StringUtils.isNotBlank(model.getName())) {
```
Comment (hasuniea) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/wsdl/WSDL11SOAPOperationExtractor.java:310]: Can you fix the formatting issue?

**Review by hasuniea — APPROVED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/wsdl/WSDL20ProcessorImpl.java (hunk 1 above)
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/wsdl/WSDL20ProcessorImpl.java:262]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10668/commits/3b65b5844c656f1e87dd5c96fdec4a4a22dd15de

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/wsdl/WSDL11SOAPOperationExtractor.java (hunk 2 above)
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/wsdl/WSDL11SOAPOperationExtractor.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10668/commits/e39d2b2ac4948ce899f5c94672903c727ef5aae3

**Review by tharikaGitHub — APPROVED** [review]

**Comment by malinthaprasan** [comment]
Please make sure all checks pass before merging

**Review by tharindu1st — APPROVED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java [hunk 3]
```diff
@@ -3540,10 +3540,11 @@ private API importSOAPToRESTAPI(InputStream fileInputStream, Attachment fileDeta
...
                 String filename = fileDetail.getContentDisposition().getFilename();
                 if (filename.endsWith(".zip")) {
-                    swaggerStr = SOAPOperationBindingUtils.getSoapOperationMapping(wsdlArchiveExtractedPath);
+                    swaggerStr = SOAPOperationBindingUtils.getSoapOperationMapping(wsdlArchiveExtractedPath, url);
                 } else if (filename.endsWith(".wsdl")) {
-                    byte[] wsdlContent = APIUtil.toByteArray(fileInputStream);
-                    swaggerStr = SOAPOperationBindingUtils.getSoapOperationMapping(wsdlContent);
+                    wsdlArchiveExtractedPath = wsdlArchiveExtractedPath.substring(0,
+                            wsdlArchiveExtractedPath.lastIndexOf('/'));
+                    swaggerStr = SOAPOperationBindingUtils.getSoapOperationMapping(wsdlArchiveExtractedPath, url);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java:3547]: Why do we pass `path` and `url` both?
In `zip` case, what does the `url` represent?
In `wsdl` case what does the file represent?
Also don't use `/` which is Linux specific path separator

Comment (hasuniea) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java:3547]: - In both approaches, the Url is set to null hence we can use the existing methods as bellow.

if (filename.endsWith(".zip")) {
                    swaggerStr = SOAPOperationBindingUtils.getSoapOperationMapping(wsdlArchiveExtractedPath);
                } else if (filename.endsWith(".wsdl")) {
                    byte[] wsdlContent = APIUtil.toByteArray(fileInputStream);
                    swaggerStr = SOAPOperationBindingUtils.getSoapOperationMapping(wsdlContent);
                } else {
                    throw new APIManagementException(ExceptionCodes.UNSUPPORTED_WSDL_FILE_EXTENSION);
                }
}

   -  **wsdlArchiveExtractedPath** is the location where we keep the extracted archive files temporally.        
   -  We can remove the slash separator and pass the inputstream

Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java:3547]: @malinthaprasan Shall I proceed with the suggested appproch?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/wsdl/util/SOAPOperationBindingUtils.java [hunk 4]
```diff
@@ -110,8 +110,24 @@ public static String getSoapOperationMapping(byte [] wsdlContent) throws APIMana
...
+     * @return swagger json string with the soap operation mapping
+     * @throws APIManagementException if an error occurs when generating swagger
+     */
+    public static String getSoapOperationMapping(String path, String url, byte [] wsdlContent)
+            throws APIManagementException {
+        APIMWSDLReader wsdlReader = new APIMWSDLReader(url);
+        WSDL11SOAPOperationExtractor processor = APIMWSDLReader.getWSDLSOAPOperationExtractor(path, wsdlReader);
         WSDLInfo wsdlInfo = processor.getWsdlInfo();
         return getGeneratedSwaggerFromWSDL(wsdlInfo);
     }
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/wsdl/util/SOAPOperationBindingUtils.java:133]: What is the usage of this method?
byte [] wsdlContent is also never used inside. I don't see any difference with this with the method which is just above.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java (hunk 3 above)
Comment (malinthaprasan) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java:3547]: So why did you need to change this?

Comment (hasuniea) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java:3547]: Not a special reason to add it. I added it without a proper understating.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/wsdl/util/SOAPOperationBindingUtils.java (hunk 4 above)
Comment (hasuniea) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/wsdl/util/SOAPOperationBindingUtils.java:133]: Yes, I never used this method. @Chamindu36 Can u remove this method?

Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/wsdl/util/SOAPOperationBindingUtils.java]: Removed with b736f4e7253bd8f35a7841654afe80b3a4acba3c

**Review by hasuniea — APPROVED** [review]

**Review by malinthaprasan — APPROVED** [review]

**Pull request merged**